tool/src/org/antlr/v5/automata/LexerATNFactory.java

```diff
@@ -351,121 +361,52 @@ public Handle charSetLiteral(GrammarAST charSetAST) {
-			this.prevCodePoint = prevCodePoint;
-			this.prevProperty = prevProperty;
-		}
-
-		@Override
-		public String toString() {
-			return String.format(
-					"%s mode=%s inRange=%s prevCodePoint=%d prevProperty=%s",
-					super.toString(),
-					mode,
-					inRange,
-					prevCodePoint,
-					prevProperty);
-		}
-
-		@Override
-		public boolean equals(Object other) {
-			if (!(other instanceof CharSetParseState)) {
-				return false;
-			}
-			CharSetParseState that = (CharSetParseState) other;
-			if (this == that) {
-				return true;
-			}
-			return Objects.equals(this.mode, that.mode) &&
-				Objects.equals(this.inRange, that.inRange) &&
-				Objects.equals(this.prevCodePoint, that.prevCodePoint) &&
-				Objects.equals(this.prevProperty, that.prevProperty);
-		}
-
-		@Override
-		public int hashCode() {
-			return Objects.hash(mode, inRange, prevCodePoint, prevProperty);
-		}
-	}
-
 	public IntervalSet getSetFromCharSetLiteral(GrammarAST charSetAST) {
 		String chars = charSetAST.getText();
-		chars = chars.substring(1, chars.length() - 1);
 		IntervalSet set = new IntervalSet();
 		CharSetParseState state = CharSetParseState.NONE;
 
-		int n = chars.length();
-		for (int i = 0; i < n; ) {
-			if (state.mode == CharSetParseState.Mode.ERROR) {
-				return new IntervalSet();
-			}
-			int c = chars.codePointAt(i);
-			int offset = Character.charCount(c);
-			if (c == '\\') {
-				EscapeSequenceParsing.Result escapeParseResult =
-					EscapeSequenceParsing.parseEscape(chars, i);
-				switch (escapeParseResult.type) {
-					case INVALID:
-						String invalid = chars.substring(escapeParseResult.startOffset,
-						                                 escapeParseResult.startOffset+escapeParseResult.parseLength);
-						g.tool.errMgr.grammarError(ErrorType.INVALID_ESCAPE_SEQUENCE,
-						                           g.fileName, charSetAST.getToken(), invalid);
-						state = CharSetParseState.ERROR;
-						break;
-					case CODE_POINT:
-						state = applyPrevStateAndMoveToCodePoint(charSetAST, set, state, escapeParseResult.codePoint);
-						break;
-					case PROPERTY:
-						state = applyPrevStateAndMoveToProperty(charSetAST, set, state, escapeParseResult.propertyIntervalSet);
-						break;
-				}
-				offset = escapeParseResult.parseLength;
-			}
-			else if (c == '-' && !state.inRange && i != 0 && i != n - 1 && state.mode != CharSetParseState.Mode.NONE) {
+		int startIndex = 1; // ignore first [
+		int endIndex = chars.length() - 1; // ignore last ]
+		int index = startIndex;
+		while (index < endIndex) {
+			if (chars.codePointAt(index) == '-' &&
+					!state.inRange && index != startIndex && index != endIndex - 1 && state.mode != CharSetParseState.Mode.NONE
+			) {
 				if (state.mode == CharSetParseState.Mode.PREV_PROPERTY) {
 					g.tool.errMgr.grammarError(ErrorType.UNICODE_PROPERTY_NOT_ALLOWED_IN_RANGE,
 							g.fileName, charSetAST.getToken(), charSetAST.getText());
-					state = CharSetParseState.ERROR;
+					return set;
```

Comment:
smells like a big change in behavior ? previous implementation returns an empty set on the next loop iteration, this returns a set possibly not empty ?

Reply:
It's not a big change in behavior because previously we had the following checks at the beginning of loop over char set and after this loop:

```java
if (state.mode == CharSetParseState.Mode.ERROR) {
	return new IntervalSet();
}
```

Now we have only one return that returns empty `IntervalSet` in case of an error instead of two ones.

The only thing that changed is about returning not empty set but partially filled in case of error. But I think it's even more correct since the error `UNICODE_PROPERTY_NOT_ALLOWED_IN_RANGE` is reported anyway and potentially invalid code generation is not performed.

Reply:
But why not return a new IntervalSet() ? Then there's no change in behavior at all ?

Reply:
I adhere the following principle: we should process code as correct as it's possible even if it contains errors. Returning partially filled char set is more correct here compare to returning empty set in case of any error. But I'm not strogly vote for that and if you insist, I can change as you suggest.